src/main/java/net/wasdev/wlp/common/plugins/util/DevUtil.java

```diff
@@ -0,0 +1,974 @@
+                }
+                
+                // check if resourceDirectory has been added
+                for (File resourceDir : resourceDirs){
+                    if (!resourceMap.get(resourceDir)) {
+                        if (resourceDir.exists()) {
+                            resourceMap.put(resourceDir, true);
+                            debug("Resource directory has been added: " + resourceDir);
+                            info("The resource directory " + resourceDir + "has been added. Restart liberty:dev mode for it to take effect.");
+                        }
+                    }
+                }
+
+                try {
+                    final WatchKey wk = watcher.poll(1, TimeUnit.SECONDS);
+                    for (WatchEvent<?> event : wk.pollEvents()) {
+                        final Path changed = (Path) event.context();
+
+                        final Watchable watchable = wk.watchable();
+                        final Path directory = (Path) watchable;
+                        debug("Processing events for watched directory: " + directory);
+
+                        File fileChanged = new File(directory.toString(), changed.toString());
+                        debug("Changed: " + changed + "; " + event.kind());
+
+                        // resource file check
+                        File resourceParent = null;
+                        for (File resourceDir : resourceDirs) {
+                            if (directory.startsWith(resourceDir.toPath())) {
+                                resourceParent = resourceDir;
+                            }
+                        }
+                        
+                        int numApplicationUpdatedMessages = countApplicationUpdatedMessages();
+
+                        // src/main/java directory
+                        if (directory.startsWith(this.sourceDirectory.toPath())) {
+                            ArrayList<File> javaFilesChanged = new ArrayList<File>();
+                            javaFilesChanged.add(fileChanged);
+                            if (fileChanged.exists() && fileChanged.getName().endsWith(".java")
+                                    && (event.kind() == StandardWatchEventKinds.ENTRY_MODIFY
+                                            || event.kind() == StandardWatchEventKinds.ENTRY_CREATE)) {
+                                debug("Java source file modified: " + fileChanged.getName());
+
+                                // tests are run in recompileJavaSource
+                                recompileJavaSource(javaFilesChanged, artifactPaths, executor, outputDirectory,
+                                        testOutputDirectory);
+                            } else if (event.kind() == StandardWatchEventKinds.ENTRY_DELETE) {
+                                debug("Java file deleted: " + fileChanged.getName());
+                                deleteJavaFile(fileChanged, outputDirectory, this.sourceDirectory);
```

Comment:
Does this recompile the Java source? After deleting a Java file, it probably should.

Reply:
In general, we only recompile the java files that were changed.  In this case, a java file was deleted so we just delete the corresponding class file.  We could potentially recompile the entire source in order to surface any compile errors if other classes reference the deleted class, but that seems excessive.  In the current implementation, what someone would see (if something still referenced the deleted file) would be an error at runtime, as well as a compile error shown by their language server in whatever IDE they are using (which should be enough for this scenario).